src/python/pants/backend/docker/lint/hadolint/rules.py

```diff
@@ -50,37 +52,36 @@ async def run_hadolint(request: HadolintRequest, hadolint: Hadolint) -> LintResu
     if hadolint.skip:
         return LintResults([], linter_name="Hadolint")
 
-    downloaded_hadolint, sources, config_files = await MultiGet(
+    downloaded_hadolint, config_files = await MultiGet(
         Get(DownloadedExternalTool, ExternalToolRequest, hadolint.get_request(Platform.current)),
-        Get(
-            SourceFiles,
-            SourceFilesRequest(
-                [field_set.source for field_set in request.field_sets],
-                for_sources_types=(DockerImageSourceField,),
-                enable_codegen=True,
-            ),
-        ),
         Get(ConfigFiles, ConfigFilesRequest, hadolint.config_request()),
     )
+
+    dockerfile_infos = await MultiGet(
+        Get(DockerfileInfo, DockerfileInfoRequest(field_set.address))
+        for field_set in request.field_sets
+    )
+
     input_digest = await Get(
         Digest,
         MergeDigests(
             (
-                sources.snapshot.digest,
                 downloaded_hadolint.digest,
                 config_files.snapshot.digest,
+                *[info.digest for info in dockerfile_infos],
```

Comment:
I think this will keep it a generator statement, rather than creating an intermediate list.

```suggestion
                *(info.digest for info in dockerfile_infos),
```